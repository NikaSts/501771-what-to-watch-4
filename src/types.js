import {string, number, shape, arrayOf, bool} from 'prop-types';

export const movieType = shape({
  id: number.isRequired,
  title: string.isRequired,
  runTime: number.isRequired,
  genre: string.isRequired,
  releaseDate: number.isRequired,
  poster: string.isRequired,
  previewImage: string.isRequired,
  backgroundImage: string.isRequired,
  backgroundColor: string.isRequired,
  ratingScore: number.isRequired,
  ratingLevel: string.isRequired,
  ratingCount: number.isRequired,
  paragraphs: arrayOf(string).isRequired,
  director: string.isRequired,
  stars: arrayOf(string).isRequired,
  preview: string.isRequired,
  isFavorite: bool.isRequired,
});

export const promoMovieType = shape({
  id: number.isRequired,
  title: string.isRequired,
  genre: string.isRequired,
  releaseDate: number.isRequired,
  poster: string.isRequired,
  backgroundImage: string.isRequired,
  backgroundColor: string.isRequired,
});

export const cardMovieType = shape({
  title: string.isRequired,
  image: string.isRequired,
});
