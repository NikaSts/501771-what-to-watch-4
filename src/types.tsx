export interface MovieType {
  id: number;
  title: string;
  runTime: number;
  genre: string;
  releaseDate: number;
  poster: string;
  previewImage: string;
  backgroundImage: string;
  backgroundColor: string;
  ratingScore: number;
  ratingLevel?: string;
  ratingCount: number;
  description: string;
  director: string;
  stars: Array<string>;
  preview: string;
  video: string;
  isFavorite: boolean;
}

export interface UserDataType {
  id: number;
  name: string;
  email: string;
  avatar: string;
}

export interface ReviewType {
  id: number;
  user: {
    id: number;
    name: string;
  };
  rating: number;
  comment: string;
  date: string;
}
