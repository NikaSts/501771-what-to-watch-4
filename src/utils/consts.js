export const Tab = {
  OVERVIEW: `Overview`,
  REVIEWS: `Reviews`,
  DETAILS: `Details`,
};

export const tabs = [Tab.OVERVIEW, Tab.REVIEWS, Tab.DETAILS];

export const DEFAULT_GENRE = `All genres`;

export const MOVIES_TO_SHOW_COUNT = 8;
export const MAX_GENRES_TO_SHOW = 10;
export const MAX_SIMILAR_MOVIES = 4;
export const MAX_ACTORS_TO_SHOW = 4;

export const Rating = {
  BAD: `Bad`,
  NORMAL: `Normal`,
  GOOD: `Good`,
  VERY_GOOD: `Very good`,
  AWESOME: `Awesome`,
};

export const Score = {
  BAD: 60,
  NORMAL: 70,
  GOOD: 80,
  VERY_GOOD: 90,
  AWESOME: 100,
};

export const DELAY = 1000;

export const ActionType = {
  OPEN_FULL_SCREEN_PLAYER: `openFullScreenPlayer`,
  CLOSE_FULL_SCREEN_PLAYER: `closeFullScreenPlayer`,
  GET_MOVIES: `getMovies`,
  GET_PROMO_MOVIE: `getPromoMovie`,
  GET_REVIEWS: `getReviews`,
  REQUIRE_AUTHORIZATION: `requireAuthorization`,
  IS_AUTHORIZING: `isAuthorizing`,
  IS_NOT_AUTHORIZING: `isNotAuthorizing`,
  IS_AUTHORIZATION_ERROR: `isAuthorizationError`,
  LOAD_USER_DATA: `loadUserData`,
};

export const AuthorizationStatus = {
  AUTH: `AUTH`,
  NO_AUTH: `NO_AUTH`,
};

export const EntryPoint = {
  REVIEWS: `/wtw/comments`,
  FAVORITE: `/wtw/favorite`,
  MOVIES: `/wtw/films`,
  PROMO: `/wtw/films/promo`,
  LOGIN: `/wtw/login`,
};

export const TIMER = 5000;

export const Error = {
  UNAUTHORIZED: 401,
};

export const URL = `https://4.react.pages.academy`;

export const AppRoute = {
  ROOT: `/`,
  LOGIN: `/login`,
  MY_LIST: `/mylist`,
  MOVIE_PAGE: `/films`,
};

export const TEXTAREA_COLOR = `rgba(255, 255, 255, 0.3)`;
