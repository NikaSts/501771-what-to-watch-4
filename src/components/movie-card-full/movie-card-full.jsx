import React from 'react';
import {oneOfType, shape, arrayOf, number, string} from 'prop-types';
import {connect} from 'react-redux';

import MovieInfo from '../movie-info/movie-info';
import PageHeader from '../page-header/page-header';
import Tabs from '../tabs/tabs';
import withActiveItem from '../../hocs/with-active-item/with-active-item';
import {movieType} from '../../types';
import {getMovies, getReviews} from '../../store/movies/selectors';
import {getAuthorizationStatus, getUserData} from '../../store/user/selectors';
import {AuthorizationStatus} from '../../utils/consts';


const TabsWithActiveItem = withActiveItem(Tabs);

const MovieCardFull = ({
  id, movies, reviews, authorizationStatus, userData
}) => {
  const activeMovie = movies.find((movie) => movie.id === id);
  const {title, genre, releaseDate, poster, backgroundImage, backgroundColor} = activeMovie;
  const isAuth = authorizationStatus === AuthorizationStatus.AUTH;

  return (
    <section
      key={id}
      className="movie-card movie-card--full"
      style={{backgroundColor: `${backgroundColor}`}}>
      <div className="movie-card__hero">
        <div className="movie-card__bg">
          <img src={backgroundImage} alt={title} />
        </div>
        <h1 className="visually-hidden">WTW</h1>

        <PageHeader
          isAuth={isAuth}
          userData={userData}
        />
        <div className="movie-card__wrap">
          <MovieInfo
            id={id}
            title={title}
            genre={genre}
            releaseDate={releaseDate}
          />
        </div>
      </div>

      <div className="movie-card__wrap movie-card__translate-top">
        <div className="movie-card__info">
          <div className="movie-card__poster movie-card__poster--big">
            <img src={poster} alt={title} width="218" height="327" />
          </div>
          <TabsWithActiveItem
            activeMovie={activeMovie}
            reviews={reviews}
          />
        </div>
      </div>
    </section>
  );
};

MovieCardFull.propTypes = {
  id: number.isRequired,
  movies: arrayOf(movieType).isRequired,
  authorizationStatus: string.isRequired,
  reviews: oneOfType([
    arrayOf(shape({
      id: number.isRequired,
      user: shape({
        id: number.isRequired,
        name: string.isRequired,
      }).isRequired,
      rating: number.isRequired,
      comment: string.isRequired,
      date: string.isRequired,
    })),
  ]),
  userData: shape({
    id: number.isRequired,
    name: string.isRequired,
    email: string.isRequired,
    avatar: string.isRequired,
  })
};

const mapStateToProps = (state) => ({
  movies: getMovies(state),
  reviews: getReviews(state),
  authorizationStatus: getAuthorizationStatus(state),
  userData: getUserData(state),
});


export {MovieCardFull};
export default connect(mapStateToProps)(MovieCardFull);
