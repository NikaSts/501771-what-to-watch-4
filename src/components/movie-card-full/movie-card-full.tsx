import * as React from 'react';
import {connect} from 'react-redux';

import MovieInfo from '../movie-info/movie-info';
import PageHeader from '../page-header/page-header';
import Tabs from '../tabs/tabs';
import withActiveItem from '../../hocs/with-active-item/with-active-item';
import {getReviews} from '../../store/movies/selectors';
import {getAuthorizationStatus, getUser} from '../../store/user/selectors';
import {Operation as MoviesOperation} from '../../store/movies/actions';
import {MovieType, UserDataType, ReviewType} from '../../types';
import {AuthorizationStatus} from '../../utils/consts';


const TabsWithActiveItem = withActiveItem(Tabs);

interface MovieCardFullProps {
  currentPage: string;
  id: number;
  activeMovie: MovieType;
  authorizationStatus: string;
  reviews: Array<ReviewType>;
  user: UserDataType;
  onIsFavoriteButtonClick: () => void;
}

const MovieCardFull: React.FC<MovieCardFullProps> = ({
  currentPage, id, activeMovie, reviews, authorizationStatus, user, onIsFavoriteButtonClick
}: MovieCardFullProps) => {
  const {
    title, genre, releaseDate, poster, backgroundImage, backgroundColor, isFavorite
  } = activeMovie;
  const isAuth: boolean = authorizationStatus === AuthorizationStatus.AUTH;

  return (
    <section
      key={id}
      className="movie-card movie-card--full"
      style={{backgroundColor: `${backgroundColor}`}}>
      <div className="movie-card__hero">
        <div className="movie-card__bg">
          <img src={backgroundImage} alt={title} />
        </div>
        <h1 className="visually-hidden">WTW</h1>

        <PageHeader
          isAuth={isAuth}
          currentPage={currentPage}
          user={user}
        />
        <div className="movie-card__wrap">
          <MovieInfo
            currentPage={currentPage}
            id={id}
            title={title}
            genre={genre}
            releaseDate={releaseDate}
            isFavorite={isFavorite}
            onIsFavoriteButtonClick={onIsFavoriteButtonClick}
            isAuth={isAuth}
          />
        </div>
      </div>

      <div className="movie-card__wrap movie-card__translate-top">
        <div className="movie-card__info">
          <div className="movie-card__poster movie-card__poster--big">
            <img src={poster} alt={title} width="218" height="327" />
          </div>
          <TabsWithActiveItem
            //           currentPage={currentPage}
            activeMovie={activeMovie}
            reviews={reviews}
          />
        </div>
      </div>
    </section>
  );
};

const mapStateToProps = (state) => ({
  reviews: getReviews(state),
  authorizationStatus: getAuthorizationStatus(state),
  user: getUser(state),
});

const mapDispatchToProps = (dispatch, props) => ({
  onIsFavoriteButtonClick() {
    const {id, isFavorite} = props.activeMovie;
    dispatch(MoviesOperation.sendFavoriteMovie(id, isFavorite));
  },
});


export {MovieCardFull};
export default connect(mapStateToProps, mapDispatchToProps)(MovieCardFull);
