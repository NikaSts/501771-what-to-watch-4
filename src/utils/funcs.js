import {DEFAULT_GENRE, Rating, Score} from './consts';

export const getRandomNumber = (min, max) => {
  return Math.floor(Math.random() * (max - min)) + min;
};

export const getRandomItem = (items) => {
  return items[getRandomNumber(0, items.length)];
};

export const getUrlFromTitle = (title) => {
  return title
    .toLowerCase()
    .split(` `)
    .join(`-`);
};

export const getRandomBoolean = () => {
  return Math.random() > 0.5;
};

export const getRandomItems = (items) => {
  const result = [];
  [...items].forEach((item) => {
    if (getRandomBoolean()) {
      result.push(item);
    }
  });
  return result;
};

export const suffleItems = (items) => {
  for (let i = items.length - 1; i > 0; i--) {
    const randomIndex = Math.floor(Math.random() * (i + 1));
    const temp = items[randomIndex];
    items[randomIndex] = items[i];
    items[i] = temp;
  }
  return items;
};

export const getRatingLevel = (score) => {
  if (score < Score.BAD) {
    return Rating.BAD;
  }
  if (score < Score.NORMAL) {
    return Rating.NORMAL;
  }
  if (score < Score.GOOD) {
    return Rating.GOOD;
  }
  if (score < Score.VERY_GOOD) {
    return Rating.VERY_GOOD;
  }
  return Rating.AWESOME;
};

export const getMoviesToShow = (movies, activeGenre) => {
  if (activeGenre === DEFAULT_GENRE) {
    return [...movies];
  }
  return [...movies].filter((movie) => movie.genre === activeGenre);
};
