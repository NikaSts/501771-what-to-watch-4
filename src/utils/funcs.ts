
import {DEFAULT_GENRE, Rating, Score} from './consts';
import {MovieType} from '../types';

interface ExtendInterface {
    <T>(state: T, newStateValue: T): T;
}

export const extend: ExtendInterface = (state, newStateValue) => {
  return Object.assign({}, state, newStateValue);
};

export const getRatingLevel = (score: number) => {
  if (score < Score.BAD) {
    return Rating.BAD;
  }
  if (score < Score.NORMAL) {
    return Rating.NORMAL;
  }
  if (score < Score.GOOD) {
    return Rating.GOOD;
  }
  if (score < Score.VERY_GOOD) {
    return Rating.VERY_GOOD;
  }
  return Rating.AWESOME;
};

export const filterMovies = (movies: Array<MovieType>, activeGenre: string = DEFAULT_GENRE) => {
  if (activeGenre === DEFAULT_GENRE) {
    return [...movies];
  }
  return [...movies].filter((movie) => movie.genre === activeGenre);
};

export const getGenresFromMovies = (movies: Array<MovieType>) => {
  return [DEFAULT_GENRE, ...new Set(movies.map((movie) => movie.genre))];
};

export const getMovieById = (movies: Array<MovieType>, id: number) => {
  return movies.find((movie) => movie.id === id);
};

export const formatTimeWithSeconds = (time: number) => {
  const minutes = Math.trunc(time / 60);
  const seconds = Math.trunc(time % 60);
  const hours = Math.trunc(minutes / 60);

  return [
    (`0` + hours).slice(-2),
    (`0` + minutes).slice(-2),
    (`0` + seconds).slice(-2)
  ].join(`:`);
};

export const formatTime = (time: number) => {
  const hours = Math.trunc(time / 60);
  const minutes = time % 60;

  if (hours) {
    return `${hours}h ${minutes}m`;
  }
  return `${minutes}m`;
};
