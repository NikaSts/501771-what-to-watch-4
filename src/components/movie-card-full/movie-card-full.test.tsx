import * as React from 'react';
import * as renderer from 'react-test-renderer';
import {Provider} from 'react-redux';
import configureStore from 'redux-mock-store';
import {Router} from 'react-router-dom';
import history from '../../history';
import {MovieCardFull} from './movie-card-full';
import {Page} from '../../utils/consts';
import {MovieType, UserDataType, ReviewType} from '../../types';

const mockStore = configureStore([]);

const activeMovie: MovieType = {
  id: 1,
  title: `Active Movie`,
  runTime: 1,
  genre: ``,
  releaseDate: 1,
  poster: ``,
  previewImage: ``,
  backgroundImage: ``,
  backgroundColor: ``,
  ratingScore: 1,
  ratingLevel: ``,
  ratingCount: 1,
  description: ``,
  director: ``,
  stars: [``],
  preview: ``,
  video: ``,
  isFavorite: true,
};
const user: UserDataType = {
  id: 1,
  name: ``,
  email: ``,
  avatar: ``,
};
const currentPage: string = Page.MOVIE_PAGE;
const reviews: Array<ReviewType> = [{
  id: 2,
  user: {
    id: 15,
    name: `Kendall`
  },
  rating: 8.7,
  comment: ``,
  date: ``
}];

const noop = () => {
  // Mock function for test props
};

it(`MovieCardFull should render correctly`, () => {
  const store = mockStore({
    USER: {
      authorizationStatus: `AUTH`,
      user,
    }
  });

  const tree = renderer
    .create(
        <Provider store={store}>
          <Router history={history}>
            <MovieCardFull
              id={1}
              currentPage={currentPage}
              authorizationStatus={`AUTH`}
              activeMovie={activeMovie}
              reviews={reviews}
              user={user}
              onIsFavoriteButtonClick={noop}
            />
          </Router>
        </Provider>, {
          createNodeMock: () => {
            return {};
          }
        })
      .toJSON();

  expect(tree).toMatchSnapshot();
});
