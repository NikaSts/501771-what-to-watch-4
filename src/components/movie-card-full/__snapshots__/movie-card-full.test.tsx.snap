// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieCardFull should render correctly 1`] = `
<section
  className="movie-card movie-card--full"
  style={
    Object {
      "backgroundColor": "",
    }
  }
>
  <div
    className="movie-card__hero"
  >
    <div
      className="movie-card__bg"
    >
      <img
        alt="Active Movie"
        src=""
      />
    </div>
    <h1
      className="visually-hidden"
    >
      WTW
    </h1>
    <header
      className="page-header"
    >
      <div
        className="logo"
      >
        <a
          className="logo__link"
          href="/"
          onClick={[Function]}
        >
          <span
            className="logo__letter logo__letter--1"
          >
            W
          </span>
          <span
            className="logo__letter logo__letter--2"
          >
            T
          </span>
          <span
            className="logo__letter logo__letter--3"
          >
            W
          </span>
        </a>
      </div>
      <div
        className="user-block"
      >
        <a
          href="/mylist"
          onClick={[Function]}
        >
          <div
            className="user-block__avatar"
          >
            <img
              alt="User avatar"
              height="63"
              src="https://4.react.pages.academy"
              width="63"
            />
          </div>
        </a>
      </div>
    </header>
    <div
      className="movie-card__wrap"
    >
      <div
        className="movie-card__desc"
      >
        <h2
          className="movie-card__title"
        >
          Active Movie
        </h2>
        <p
          className="movie-card__meta"
        >
          <span
            className="movie-card__genre"
          >
            
          </span>
          <span
            className="movie-card__year"
          >
            1
          </span>
        </p>
        <div
          className="movie-card__buttons"
        >
          <a
            className="btn btn--play movie-card__button"
            href="/player/1"
            onClick={[Function]}
            type="button"
          >
            <svg
              height="19"
              viewBox="0 0 19 19"
              width="19"
            >
              <use
                xlinkHref="#play-s"
              />
            </svg>
            <span>
              Play
            </span>
          </a>
          <button
            className="btn btn--list movie-card__button"
            onClick={[Function]}
            type="button"
          >
            <svg
              height="14"
              viewBox="0 0 18 14"
              width="18"
            >
              <use
                xlinkHref="#in-list"
              />
            </svg>
            <span>
              My list
            </span>
          </button>
          <a
            className="btn btn--review movie-card__button"
            href="/films/1/review"
            onClick={[Function]}
          >
            Add review
          </a>
        </div>
      </div>
    </div>
  </div>
  <div
    className="movie-card__wrap movie-card__translate-top"
  >
    <div
      className="movie-card__info"
    >
      <div
        className="movie-card__poster movie-card__poster--big"
      >
        <img
          alt="Active Movie"
          height="327"
          src=""
          width="218"
        />
      </div>
      <div
        className="movie-card__desc"
      >
        <nav
          className="movie-nav movie-card__nav"
        >
          <ul
            className="movie-nav__list"
          >
            <li
              className="movie-nav__item movie-nav__item--active"
            >
              <a
                className="movie-nav__link"
                href="#"
                onClick={[Function]}
              >
                Overview
              </a>
            </li>
            <li
              className="movie-nav__item"
            >
              <a
                className="movie-nav__link"
                href="#"
                onClick={[Function]}
              >
                Reviews
              </a>
            </li>
            <li
              className="movie-nav__item"
            >
              <a
                className="movie-nav__link"
                href="#"
                onClick={[Function]}
              >
                Details
              </a>
            </li>
          </ul>
        </nav>
        <div
          className="movie-rating"
        >
          <div
            className="movie-rating__score"
          >
            1
          </div>
          <p
            className="movie-rating__meta"
          >
            <span
              className="movie-rating__level"
            >
              Bad
            </span>
            <span
              className="movie-rating__count"
            >
              1
               ratings
            </span>
          </p>
        </div>
        <div
          className="movie-card__text"
        >
          
          <p
            className="movie-card__director"
          >
            <strong>
              Director: 
              
            </strong>
          </p>
          <p
            className="movie-card__starring"
          >
            <strong>
              Starring: 
              
               and other
            </strong>
          </p>
        </div>
      </div>
    </div>
  </div>
</section>
`;
